HomeWork | Budlyanskiy Roman

romutchio opened the pull request:
Mentor: @BurningMarshmallow

romutchio committed:
Add TagCloud project;

romutchio committed:
IExceptonHandler + MessageBoxHandler;
Solution file management;

romutchio committed:
Error Handling via Result<T>;
File Management in project;

BurningMarshmallow commented:
Давай сначала наведём косметический порядок :).
Во-первых, ко всем файлам можно применить форматирование, чтобы пропали лишние пустые строки/появились пробелы в нужных местах для лучшей читаемости.

Во-вторых, нужно убрать весь неиспользуемый код. Его совсем чуть-чуть (неиспользуемые юзинги и несколько переменных(к примеру center, layout в тестах) + классов, а еще есть закоменченый код(или там есть что-то важное?).

В-третьих, есть менее серъезные замечания по Code Style, которые подсказывает Resharper: не у всех классов расставлены модификаторы доступа(public/private) + namespace does not correspond to file location, к примеру TagCloud.Interfaces у ICloudLayouter.

BurningMarshmallow commented on TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System;
```
Это версия программы с предыдущих блоков? Давай уберём её из репо, у нас ведь теперь есть новая, крутая, с GUI :)

BurningMarshmallow commented on TagCloud/Filters & Preprocessors/BasicFormConverter.cs:
```diff
@@ -0,0 +1,24 @@
+using System.Collections.Generic;
```
Кстати, почему Filters & Preprocessors лежат в папке в корне репо? По смыслу тут же происходит работа только со словами, так что её можно переместить в Words. Заодно и интерфейс IWordFilter будет поближе ;)

BurningMarshmallow commented on TagCloud/Actions/SaveImageAction.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public void Perform()
+        {
+            using (OpenFileDialog openFileDialog = new OpenFileDialog())
```
Правда ли, что OpenFileDialog не самый идеальный вариант для этого действия? Может есть что-то специально для сохранения?

BurningMarshmallow commented on TagCloud/Actions/LoadExcludingWordsAction.cs:
```diff
@@ -0,0 +1,45 @@
+            this.reader = reader;
+            this.exceptionHandler = exceptionHandler;
+        }
+        public string Category { get; } = "File";
```
Вкусовщина: у тебя где-то свойства описываются через  { get; } =  , где-то через =>. Давай везде напишем через =>, так поприятнее выглядит, как думаешь?

romutchio committed:
Code Cleanup + Reformat;

romutchio committed:
Removed old version of TagCloudVisualization;

BurningMarshmallow commented on TagCloud/Words/IRepository.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace TagCloud.Words
+{
+    public interface IRepository
```
Неплохо, что IRepository выглядит такой обобщенной абстракцией, но кое-что его выдает (IEnumerable<string> words). Здесь можно выбрать из 2 вариантов: либо признать, что это именно IWordRepository, либо забыть, что это называется words(в классе тогда тоже будет обобщенное название, но кажется, из контекста будет понятно, что это WordRepository).

BurningMarshmallow commented:
Следующим шагом можно вспомнить про пользователя GUI: во-первых, некоторые подсказки и названия тяжело читать(к примеру, Load wordsRepository to exclude from Tag Cloud), во-вторых, можно улучшить всплывашку об ошибке(давай добавим к ней заголовок и сделаем её прям ошибкой, чтобы был красный цвет и пользователь мог быстрее понять, что что-то пошло не так), в-третьих, можно скрыть консоль, которая появляется при запуске(она никому не нужна).

BurningMarshmallow commented:
А еще можно пройтись по проекту и поискать опечатки(к примеру, сейчас есть BackdgoundColor)

romutchio committed:
Removed console when the app is running;
Better hints;
Better exception MessageBox (Red cross and title);
Some grammar correction;
Renamed IRepository -> IWordRepository;

BurningMarshmallow commented:
Теперь давай посмотрим, что написано в домашке https://github.com/kontur-courses/fp/blob/master/HomeExercise.md(воспримем её как ТЗ), и попробуем добавить фичи, чтобы все работало как описано.
Конкретнее:
1. "файл настроек не получилось прочитать" - давай перенесём наши настройки в файл(note: формат файла может быть любой, не обязательно .txt).
2. "одна из внешних библиотек дала сбой (например, NHunspell не смог найти свои словари)" - сейчас если он не найдет словарей, будет некрасивый ошибос, я проверил :).
3. "облако тегов не влезло на изображение заданного размера" - сейчас если облако не влезает, то оно все равно отрисовывается за границами окна, ведь так?

romutchio committed:
NHunspell exceptions handling;
Handling of situation, when TagCloud doesn't fit given sizes;

romutchio committed:
Settings load via *settings.ini* file;

romutchio committed:
New way of saving settings via Properties.settings;

romutchio committed:
Better Exception message;
New default values for picture (800x800);

romutchio committed:
"Hot" Settings in runtime app;

BurningMarshmallow commented on TagCloud/Settings/SettingsChecker.cs:
```diff
@@ -0,0 +1,49 @@
+        public void CreateFileWatcher(string path)
+        {
+            ReadJson($"{path}\\settings.json");
+            FileSystemWatcher watcher = new FileSystemWatcher();
```
Ага, FileSystemWatcher крутая штука, var только поставь :).

BurningMarshmallow commented on TagCloud/Settings/SettingsChecker.cs:
```diff
@@ -0,0 +1,49 @@
+
+        private void OnChanged(object source, FileSystemEventArgs e)
+        {
+            MessageBox.Show("File: " + e.FullPath + " " + e.ChangeType);
```
Сообщение можно не показывать каждый раз: на этапе разработки это конечно полезно, чтобы убедиться, что Watcher правильно настроен, но сейчас это только добавляет дополнительное действие(я даже в первый раз не понял, почему у меня настройки применились только после того, как я нажал Ok в MessageBox-е).

BurningMarshmallow commented on TagCloud/Settings/SettingsChecker.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace TagCloud.Settings
+{
+    public class SettingsChecker
```
Вкусовщина: можно переименовать этот класс в SettingsLoader или Repository, так как он не только проверяет изменение настроек, но еще и умеет возвращать актуальное состояние

BurningMarshmallow commented on TagCloud/Settings/ImageSettings.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class ImageSettings
+    {
+        public int Width { get; set; } = 800;
```
Эти дефолтные значения лучше синхронизировать с теми, что в файле :)

BurningMarshmallow commented on TagCloud/TagCloud.csproj:
```diff
@@ -0,0 +1,138 @@
+      <HintPath>..\packages\Autofac.4.9.0-beta1\lib\net45\Autofac.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="INIFileParser, Version=2.5.2.0, Culture=neutral, PublicKeyToken=79af7b307b65cf3c">
```
Важное правило: неиспользуемые пакеты/ссылки, как неиспользуемый код - за ними нужно следить и убирать.

BurningMarshmallow commented on TagsCloudVisualization_Tests/CircularCloudVisualizer_Should.cs:
```diff
@@ -0,0 +1,50 @@
+
+
+        [Test]
+        public void ShiftRectangleToCenter_ShiftSingleRectangle_CorrectShift()
```
Тест красный, насколько я вижу из-за того, что ImageSettings изменились

BurningMarshmallow commented:
Еще вроде пропала проверка на то, влезает ли облако тегов в изображение, когда мы обновляем настройки в горячем режиме. Можем ли мы это вернуть?

romutchio committed:
Bug Fixes and Code Refactoring;

romutchio committed:
Better way to Update Settings;

romutchio committed:
Better way to Update Settings;

romutchio committed:
Code Cleanup;
